### PR TITLE
[IMC] Sync Mutex Cache

### PR DESCRIPTION
**Pull request by just-in-chang**
## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: crates/aptos-in-memory-cache/src/caches/sync_mutex.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+{
+    fn get(&self, key: &usize) -> Option<SizedCacheEntry<T>> {
+        let arc = self.cache[*key % self.capacity].clone();
+        let lock = arc.lock();
+        (*lock).clone()
+    }
+
+    fn insert_with_size(&self, key: usize, value: T, size_in_bytes: usize) -> usize {
+        // Get lock for cache entry
+        let arc = self.cache[key % self.capacity].clone();
```
Comment (CapCap) [inline crates/aptos-in-memory-cache/src/caches/sync_mutex.rs:55]: let index = key % self.capacity;
        let arc = self.cache[index].clone();

File: crates/aptos-in-memory-cache/src/caches/sync_mutex.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        // Update cache entry
+        self.size.fetch_add(size_in_bytes, Ordering::Relaxed);
+        *lock = Some(SizedCacheEntry {
+            key,
+            value,
+            size_in_bytes,
+        });
+
+        return key % self.capacity;
```
Comment (CapCap) [inline crates/aptos-in-memory-cache/src/caches/sync_mutex.rs:72]: `index`

File: crates/aptos-in-memory-cache/src/caches/sync_mutex.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        let mut lock = arc.lock();
+
+        // Update cache size
+        if let Some(prev_value) = &*lock {
+            self.size
+                .fetch_sub(prev_value.size_in_bytes, Ordering::Relaxed);
+        }
+
+        // Set cache entry to None
+        *lock = None;
```
Comment (CapCap) [inline crates/aptos-in-memory-cache/src/caches/sync_mutex.rs:87]: can simplify to 
```
        // Update cache size & set previous value to none
        if let Some(prev_value) = lock.take() {
            self.size
                .fetch_sub(prev_value.size_in_bytes, Ordering::Relaxed);
        }
  ```

File: crates/aptos-in-memory-cache/src/caches/sync_mutex.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    fn total_size(&self) -> usize {
+        self.size.load(Ordering::Relaxed)
+    }
```
Comment (CapCap) [inline crates/aptos-in-memory-cache/src/caches/sync_mutex.rs:92]: what is this used for?

File: crates/aptos-in-memory-cache/tests/test.rs [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+}
+
+/// As discussed with Max previously, this cache uses a primitive Rust array as a circular buffer to store cache entries.
+/// The metadata contains three pointers: tail, head, and watermark.
+/// The actual contents of the cache are between the head and tail pointers.
+/// We guarantee that all values between the head and tail pointers exist and are valid cache entries.
+/// The watermark represents the highest key that has been inserted into the cache.
+/// It exists to enable out of order insertions because we guarantee that new values between the tail and the watermark are not evicted.
+/// As new values are inserted, the tail pointer is updated accordingly to ensure that the contents of the cache (between the tail and head pointers) are contiguous.
+#[derive(Debug, Clone)]
```
Comment (CapCap) [inline crates/aptos-in-memory-cache/tests/test.rs:28]: technically... you don't need any of these 😬

File: crates/aptos-in-memory-cache/src/caches/sync_mutex.rs (hunk 4 above)
Comment (just-in-chang) [reply crates/aptos-in-memory-cache/src/caches/sync_mutex.rs:92]: Just API to expose total size of cache for wrapper struct to use 👍

**Pull request closed without merging**